aiida_quantumespresso/calculations/pp.py

```diff
@@ -1,27 +1,240 @@
+            message='The stdout output file could not be read.')
+        spec.exit_code(411, 'ERROR_OUTPUT_DATAFILE_NOT_READ',
+            message='The formatted data output file could not be read.')
+        spec.exit_code(420, 'ERROR_OUTPUT_STDOUT_INCOMPLETE',
+            message='The stdout output file was incomplete.')
+        spec.exit_code(430, 'ERROR_UNSUPPORTED_DATAFILE_FORMAT',
+            message='The data file format is not supported by the parser')
+        spec.exit_code(500, 'ERROR_UNEXPECTED_PARSER_EXCEPTION',
+            message='The parser raised an unexpected exception.')
+
+    def prepare_for_submission(self, folder):
+        """Create the input files from the input nodes passed to this instance of the `CalcJob`.
+
+        :param folder: an `aiida.common.folders.Folder` to temporarily write files on disk
+        :return: `aiida.common.datastructures.CalcInfo` instance
+        """
+        if 'settings' in self.inputs:
+            settings = _uppercase_dict(self.inputs.settings.get_dict(), dict_name='settings')
+        else:
+            settings = {}
+
+        # Put the first-level keys as uppercase (i.e., namelist and card names) and the second-level keys as lowercase
+        if 'parameters' in self.inputs:
+            parameters = _uppercase_dict(self.inputs.parameters.get_dict(), dict_name='parameters')
+            parameters = {k: _lowercase_dict(v, dict_name=k) for k, v in six.iteritems(parameters)}
+        else:
+            parameters = {}
+
+        # set default values. NOTE: this is different from PW/CP
+        for blocked in self._blocked_keywords:
+            namelist = blocked[0].upper()
+            key = blocked[1].lower()
+            value = blocked[2]
+
+            if namelist in parameters:
+                if key in parameters[namelist]:
+                    raise exceptions.InputValidationError(
+                        "You cannot specify explicitly the '{}' key in the '{}' "
+                        'namelist.'.format(key, namelist))
+            else:
+                parameters[namelist] = {}
+            parameters[namelist][key] = value
+
+        # Set the values that have placeholders:
+        parameters['INPUTPP']['plot_num'] = self.inputs.plot_number.value
+        dimension = self.inputs.plot_dimension.value
+        parameters['PLOT']['iflag'] = dimension
+        # Restrict the plot output to the file types that we want to be able to parse
+        if dimension == 0 or dimension == 1:
+            parameters['PLOT']['output_format'] = 0    # Gnuplot, 1D
+        elif dimension == 2 or dimension == 4:
+            parameters['PLOT']['output_format'] = 7    # Gnuplot, 2D
+        elif dimension == 3:
+            parameters['PLOT']['output_format'] = 6    # Gaussian cube, 3D
+        else:
+            raise exceptions.InputValidationError(
+                "Unsupported value, {}, for 'plot_number' ('iflag' in pp.x docs)".format(dimension)
+            )
```

Comment:
if we keep `dimension` as a separate input, it is better to put the validation on the port itself

Reply:
Dropped the separate input.